### PR TITLE
TypeScript support

### PR DESCRIPTION
**Pull request by joe-re**
Many Vue projects have TypeScript declarations inside itself and support TypeScript.
[vue-router](https://github.com/vuejs/vue-router), [vuex](https://github.com/vuejs/vuex), etc...

I want to support TypeScript on our project too.

I implement typescript declarations and run typed test for [this example](https://github.com/Akryum/vue-apollo-example).
(added `yarn run test:types` command)

I thought I checked all the APIs, but please let me know if there are missing parts.

### CODE REVIEW

**Comment by joe-re** [comment]
ref: #54

**Comment by Akryum** [comment]
Thanks a lot for your PR! I'll check it out when I have some time. 😉

**Comment by salticus** [comment]
@joe-re, thank you!
For the impatient I put these changes in a repository. `yarn add  https://github.com/salticus/types-vue-apollo#v0.0.1` will hopefully take care of satisfying typescript for vue-apollo (it has for us so far.)

File: types/vue-apollo.d.ts [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import { subscribe } from 'graphql/subscription/subscribe';
+
+// include Omit type from https://github.com/Microsoft/TypeScript/issues/12215
+type Diff<T extends string, U extends string> = ({ [P in T]: P } & { [P in U]: never } & { [x: string]: never })[T];
+type Omit<T, K extends keyof T> = { [P in Diff<keyof T, K>]?: T[P] };
+
+export class VueApollo implements PluginObject<{}> {
+  [key: string]: any;
+  install: PluginFunction<{}>;
+  constructor (options: {defaultClient: ApolloClient<{}>});
```
Comment (ozguruysal) [inline types/vue-apollo.d.ts]: I guess we also need `defaultOptions` key here.

Comment (joe-re) [reply types/vue-apollo.d.ts]: Oh, I missed it. Is this right? 230078a

Comment (ozguruysal) [reply types/vue-apollo.d.ts]: I'm just starting to work with Vue Apollo and I'm not really sure what all `defaultOptions` are. I can't see them in the documentaion. Maybe @Akryum can confirm that.

On the other hand looks like there are more constructor options as mentioned in the docs [here](https://github.com/Akryum/vue-apollo/tree/next#constructor) and please see the source code [here](https://github.com/Akryum/vue-apollo/blob/next/src/apollo-provider.js#L6)

By the way, great work putting this together! 👍

Comment (Akryum) [reply types/vue-apollo.d.ts]: https://github.com/Akryum/vue-apollo#special-options

**Comment by ScreamZ** [comment]
Just, can you give some advice on how to trigger this ?
My vue.extends still doesn't recognize the apollo key...

**Pull request merged**